taichi/transforms/simplify.cpp

```diff
@@ -1,9 +1,207 @@
+class LocalLoadSearcher : public BasicStmtVisitor {
+ private:
+  Stmt *var;
+  bool result;
+
+ public:
+  using BasicStmtVisitor::visit;
+
+  explicit LocalLoadSearcher(Stmt *var) : var(var), result(false) {
+    TI_ASSERT(var->is<AllocaStmt>());
+    allow_undefined_visitor = true;
+    invoke_default_visitor = true;
+  }
+
+  void visit(LocalLoadStmt *stmt) override {
+    if (stmt->has_source(var)) {
+      result = true;
+    }
+  }
+
+  void visit(AtomicOpStmt *stmt) override {
+    if (stmt->dest == var) {
+      result = true;
+    }
+    // current store: $d
+    // $a = alloca
+    // $b : local store [$a <- v1]  <-- prev lstore |bstmt_|
+    // $c = atomic add($a, v2)      <-- cannot eliminate $b
+    // $d : local store [$a <- v3]
+
+    // current store: $b
+    // $a = alloca
+    // $b : local store [$a <- v1]
+    // $c = atomic add($a, v2)      <-- cannot eliminate $b
+  }
+
+  static bool run(IRNode *root, Stmt *var) {
+    LocalLoadSearcher searcher(var);
+    root->accept(&searcher);
+    return searcher.result;
+  }
+};
+
+// Find if there is a store (or AtomicOpStmt).
+class LocalStoreSearcher : public BasicStmtVisitor {
+ private:
+  const std::vector<Stmt *> &vars;
+  bool result;
+
+ public:
+  using BasicStmtVisitor::visit;
+
+  explicit LocalStoreSearcher(const std::vector<Stmt *> &vars)
+      : vars(vars), result(false) {
+    for (auto var : vars) {
+      TI_ASSERT(var->is<AllocaStmt>());
+    }
+    allow_undefined_visitor = true;
+    invoke_default_visitor = true;
+  }
+
+  void visit(LocalStoreStmt *stmt) override {
+    for (auto var : vars) {
+      if (stmt->ptr == var) {
+        result = true;
+        break;
+      }
+    }
+  }
+
+  void visit(AtomicOpStmt *stmt) override {
+    for (auto var : vars) {
+      if (stmt->dest == var) {
+        result = true;
+        break;
+      }
+    }
+  }
+
+  static bool run(IRNode *root, const std::vector<Stmt *> &vars) {
+    LocalStoreSearcher searcher(vars);
+    root->accept(&searcher);
+    return searcher.result;
+  }
+};
+
+// Find the **last** store, or return invalid if there is an AtomicOpStmt
+// after the last store.
+class LocalStoreForwarder : public BasicStmtVisitor {
+ private:
+  Stmt *var;
+  bool is_valid;
+  Stmt *result;
+
+ public:
+  using BasicStmtVisitor::visit;
+
+  explicit LocalStoreForwarder(Stmt *var)
+      : var(var), is_valid(true), result(nullptr) {
+    TI_ASSERT(var->is<AllocaStmt>());
+    allow_undefined_visitor = true;
+    invoke_default_visitor = true;
+  }
+
+  void visit(LocalStoreStmt *stmt) override {
+    if (stmt->ptr == var) {
+      is_valid = true;
+      result = stmt;
+    }
+  }
+
+  void visit(AllocaStmt *stmt) override {
+    if (stmt == var) {
+      is_valid = true;
+      result = stmt;
+    }
+  }
+
+  void visit(AtomicOpStmt *stmt) override {
+    if (stmt->dest == var) {
+      is_valid = false;
+    }
+  }
+
+  // Only if **both** branches finally store the variable with exactly the same
```

Comment:
One more optimization idea: if you find common statements in both branches, sometimes you can move them outside the `if` stmt.

Reply:
Yes. Now I want to move only the first and the last statements outside, as discussed at https://github.com/taichi-dev/taichi/issues/656#issuecomment-604642987.